File: DSA/Java/src/dsa/lib/math/ds/Anagram.java
package dsa.lib.math.ds;

import java.util.HashMap;


/**
 * Rule: Hash Table
 * Map: {letter: count_of_the_letter}. key: letter, value: count_of_the_letter
 * letter: array index, count_of_the_letter: array value
 *
 * 1. get a bucket
 * 2. produce to the bucket
 * 3. consume from the bucket
 * 4. check if the bucket is empty
 */
//LeetCode242
public class Anagram {
    public boolean isAnagram(String s, String t) {
        return isAnagram1(s, t);
    }

    public boolean isAnagram1(String s, String t) {
        int[] mapAlphabet = new int[26];
        for (char c : s.toCharArray()) mapAlphabet[hashFunc(c)]++;
        for (char c : t.toCharArray()) mapAlphabet[hashFunc(c)]--;
        for (int v : mapAlphabet) if (v != 0) return false;
        return true;
    }

    private int hashFunc(char c) {
        return c - 'a';
    }

    public boolean isAnagram11(String s, String t) {
        HashMap<Character, Integer> mapAlphabet = new HashMap<>();
        for (char c : s.toCharArray()) mapAlphabet.put(c, mapAlphabet.getOrDefault(c, 0) + 1);
        for (char c : t.toCharArray()) mapAlphabet.put(c, mapAlphabet.getOrDefault(c, 0) - 1);
        for (int v : mapAlphabet.values()) if (v != 0) return false;
        return true;
    }
}

File: DSA/Python/src/dsa/lib/math/ds/anagram.py
# LeetCode242
class Anagram:
    """
    Rule: Hash Table
    map: {letter: count_of_the_letter}
    array index: letter(hash func), array value: count of occurrences of the letter in the string
    """

    def isAnagram(self, s: str, t: str) -> bool:
        return self.isAnagram1(s, t)

    def isAnagram1(self, s: str, t: str) -> bool:
        def hash_func(c) -> int:
            return ord(c) - ord('a')

        map_alphabet = [0] * 26
        for c in s: map_alphabet[hash_func(c)] += 1
        for c in t: map_alphabet[hash_func(c)] -= 1
        for v in map_alphabet:
            if v: return False
        return True

    def isAnagram11(self, s: str, t: str) -> bool:
        def hash_func(c) -> int:
            return ord(c) - ord('a')

        map_alphabet = [0] * 26
        for c in s: map_alphabet[hash_func(c)] += 1
        for c in t: map_alphabet[hash_func(c)] -= 1
        return True if all(map(lambda v: v == 0, map_alphabet)) else False

    def isAnagram2(self, s: str, t: str) -> bool:
        map_alphabet = {}
        for c in s: map_alphabet[c] = map_alphabet.get(c, 0) + 1
        for c in t: map_alphabet[c] = map_alphabet.get(c, 0) - 1
        for v in map_alphabet.values():
            if v: return False
        return True

    def isAnagram21(self, s: str, t: str) -> bool:
        map_alphabet = {}
        for c in s: map_alphabet[c] = map_alphabet.get(c, 0) + 1
        for c in t: map_alphabet[c] = map_alphabet.get(c, 0) - 1
        return True if all(map(lambda v: v == 0, map_alphabet.values())) else False
